File: apps/api/internal/platform/platform.go
package platform

import (
	"context"
	"time"

	"connectrpc.com/connect"
	platformv1 "github.com/cosimosi/api/internal/gen/cosimosi/platform/v1"
	"google.golang.org/protobuf/types/known/timestamppb"
)

type PlatformService struct {
	now func() time.Time
}

func NewPlatformService(now func() time.Time) PlatformService {
	return PlatformService{now: now}
}

func (s PlatformService) Ping(
	ctx context.Context,
	_ *connect.Request[platformv1.PingRequest],
) (*connect.Response[platformv1.PingResponse], error) {
	now := time.Now
	if s.now != nil {
		now = s.now
	}
	return connect.NewResponse(&platformv1.PingResponse{
		Message:    "pong",
		ServerTime: timestamppb.New(now().UTC()),
		RequestId:  RequestIDFromContext(ctx),
	}), nil
}

File: packages/api-client/src/platform.ts
import { timestampFromDate } from '@bufbuild/protobuf/wkt'
import { createClient, createRouterTransport, type Client, type Interceptor, type Transport } from '@connectrpc/connect'
import { createConnectQueryKey, createQueryOptions } from '@connectrpc/connect-query-core'
import { createConnectTransport } from '@connectrpc/connect-web'

import { PlatformService } from './gen/cosimosi/platform/v1/platform_pb.ts'

export { PlatformService } from './gen/cosimosi/platform/v1/platform_pb.ts'
export type { PingRequest, PingResponse } from './gen/cosimosi/platform/v1/platform_pb.ts'
export type { Transport as ApiTransport } from '@connectrpc/connect'
export { timestampDate } from '@bufbuild/protobuf/wkt'

export interface ApiAuthTokenProvider {
  getAccessToken(): Promise<string | null>
}

export interface ApiTransportOptions {
  baseUrl: string
  useHttpGet?: boolean
  auth?: ApiAuthTokenProvider
  interceptors?: Interceptor[]
}

export function createApiTransport({ baseUrl, useHttpGet = true, auth, interceptors = [] }: ApiTransportOptions): Transport {
  return createConnectTransport({
    baseUrl,
    useHttpGet,
    interceptors: auth ? [createApiAuthInterceptor(auth), ...interceptors] : interceptors,
  })
}

export function createApiAuthInterceptor(auth: ApiAuthTokenProvider): Interceptor {
  return (next) => async (req) => {
    let token: string | null = null
    try {
      token = await auth.getAccessToken()
    } catch {
      token = null
    }
    if (token) {
      req.header.set('Authorization', `Bearer ${token}`)
    }
    return next(req)
  }
}

export function createPlatformClient(transport: Transport): Client<typeof PlatformService> {
  return createClient(PlatformService, transport)
}

export function createPlatformMockTransport(
  ping: () => { message: string; requestId?: string; serverTime?: Date },
): Transport {
  return createRouterTransport(({ service }) => {
    service(PlatformService, {
      ping() {
        const response = ping()
        const serverTime = response.serverTime ? timestampFromDate(response.serverTime) : undefined
        return {
          message: response.message,
          requestId: response.requestId ?? 'test-request-id',
          serverTime,
        }
      },
    })
  })
}

export function createPlatformServiceQueryKey(transport?: Transport) {
  return createConnectQueryKey({
    schema: PlatformService,
    transport,
    cardinality: undefined,
  })
}

export function createPlatformPingQueryKey(transport?: Transport) {
  return createConnectQueryKey({
    schema: PlatformService.method.ping,
    input: {},
    transport,
    cardinality: 'finite',
  })
}

export function createPlatformPingQueryOptions(transport: Transport) {
  return createQueryOptions(PlatformService.method.ping, undefined, { transport })
}
